lib/Reader/readerir.cpp

```diff
@@ -2020,6 +2020,91 @@ bool GenIR::isUnmanagedPointerType(llvm::Type *Type) {
   return Type->isPointerTy() && !isManagedPointerType(Type);
 }
 
+bool GenIR::isArrayType(llvm::Type *ArrayTy) {
+  // Do some basic sanity checks that this type is one we created to model
+  // a CLR array. Note we can't be 100% sure without keeping a whitelist
+  // when we create these types.
+  assert(isManagedPointerType(ArrayTy) && "expected managed pointer");
+  Type *Type = cast<PointerType>(ArrayTy)->getPointerElementType();
+  if (!Type->isStructTy()) {
+    return false;
+  }
+
+  // An array type may have varying fields depending on pointer size.
+  // Array payload is the last field and is a zero-element LLVM array.
+  unsigned int IndexOfElements = getPointerByteSize() == 4 ? 2 : 3;
+  StructType *ArrayStructType = cast<StructType>(Type);
+  unsigned ElementCount = ArrayStructType->getNumElements();
+  if (ElementCount != (IndexOfElements + 1)) {
+    return false;
+  }
+  llvm::Type *ElementsArrayFieldType =
+      ArrayStructType->getContainedType(IndexOfElements);
+  if (!ElementsArrayFieldType->isArrayTy()) {
+    return false;
+  }
+  ArrayType *ElementsArrayType = cast<ArrayType>(ElementsArrayFieldType);
+  if (!ElementsArrayType->getArrayNumElements() == 0) {
+    return false;
+  }
+
+  return true;
+}
+
+PointerType *GenIR::getArrayOfReferenceType() {
+  if (ArrayOfReferenceType == nullptr) {
+    createArrayOfReferenceType();
+  }
+  assert((ArrayOfReferenceType != nullptr) && "should be initialized");
+  return ArrayOfReferenceType;
+}
+
+void GenIR::createArrayOfReferenceType() {
```

Comment:
I thought about it but it seemed like it would be messy to try and common the code. Do you have some ideas about how this could be done?